### ISSUE TITLE
transform-runtime is not removing regenerator-runtime import added by preset-env

### ISSUE THREAD

**Issue by slavafomin** [babel/babel#10759]
Hello!

Thank you for this great tool!

# Problem

However, I'm looking at situation when the `transform-runtime` plugin is not removing global `regenerator-runtime/runtime` import, which is introduced by the `preset-env` preset.

# Use case

We were working on some libraries and were using `preset-env` to actually transpile the code to the required target execution environment as well as to add needed global `core-js` and `regenerator-runtime` imports (usage mode). We were also using `transform-runtime` to optimize babel helper imports.

And now we would like to migrate from using global `regeneratorRuntime` usages to the explicitly imported ones (pure), so we've enabled `regenerator` option of the `transform-runtime`.

# More details

## Source code

```javascript
// src/main.js

export async function foo() {
  return await new Promise(resolve => setTimeout(resolve, 1000));
}
```

## Babel config

```json
// .babelrc

{
  "presets": [
    [
      "@babel/preset-env",
      {
        "targets": "IE 11",
        "useBuiltIns": "usage",
        "corejs": 3,
        "modules": false,
        "debug": true
      }
    ]
  ],
  "plugins": [
    [
      "@babel/plugin-transform-runtime"
    ]
  ],
  "exclude": "node_modules/**"
}
```

## Transpiled result

```javascript
// dist/main.js

import "core-js/modules/es.object.to-string";
import "core-js/modules/es.promise";

import "regenerator-runtime/runtime";
import _regeneratorRuntime from "@babel/runtime/regenerator";

export function foo() {
  return _regeneratorRuntime.async(function foo$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          _context.next = 2;
          return _regeneratorRuntime.awrap(new Promise(function (resolve) {
            return setTimeout(resolve, 1000);
          }));

        case 2:
          return _context.abrupt("return", _context.sent);

        case 3:
        case "end":
          return _context.stop();
      }
    }
  });
}
```

As you can see, the generated file contains both pure and global imports for `regenerator-runtime`, however, only explicit (pure) import should be left.

---

I believe, that `transform-runtime` should remove such global imports (isn't it it's purpose in the first place?), or at least there should be an option for the `preset-env` to skip adding the global import for `regenerator-runtime` (I couldn't find any).

1. What do you think about this issue?

2. Do you know of any workarounds I can use right now to actually remove such superfluous imports?

Thank you.

**Comment by slavafomin** [comment]
The only way to work around this that I've found so far is to use [filter-imports](https://github.com/ember-cli/babel-plugin-filter-imports) plugin:

```javascript
[
  'filter-imports', {
    imports: {
      'regenerator-runtime/runtime': true,
    },
  },
],
```

**Comment by nicolo-ribaudo** [comment]
Out of curiosity, why are you ok with `core-js` using global imports but not `regenerator-runtime`?

**Comment by slavafomin** [comment]
@nicolo-ribaudo I can't get into much details right now, but we are working in complex environment with a lot of library code, end-user applications and dedicated service for polyfills and we are gradually switching between different approaches and we don't want to introduce breaking changes for application authors, so we've decided to encapsulate the `regenerator-runtime` in our libraries for now, but leave the global `core-js` imports. In our next step we would get rid of global imports for `core-js` as well.

**Comment by JMarkoski** [comment]
Another solution is to switch from `useBuiltIns: 'usage'` to `useBuiltIns: 'entry'`, since you are okay with global `core-js` pollution. You can add only `import 'core-js'` in the entry point, and leave the `regenerator: true` option on the `@babel/transform-runtime`.

**Comment by slavafomin** [comment]
@JMarkoski you should mention that this would lead to a less efficient bundles due to the addition of extraneous polyfills, that are not actually used by the library. However, it could work in certain scenarios.

**Comment by JMarkoski** [comment]
No, it won't. Adding import to `core-js` will pollute the global environment, thus adding functionality, but not increasing bundle size. Also the single line import will convert to more granular imports from the `core-js` library based on your targets, thus not adding polyfills that are not needed, although this is irrelevant. The `useBuiltIns: 'usage'` option also pollutes the global environment, but it adds polyfills based on whether you use a feature that is not supported on your targets in your code or not (this is problematic if some of the dependencies use a feature that you don't use in your code, because it won't be polyfilled)

**Comment by dangreen** [comment]
Hi. I have same issue. Using transform-runtime I expect what my bundle will dependence only on core-js and babel/runtime. I have the project which use Rollup for bundling. In rollup config I should provide external dependencies, and I provide only core-js and babel/runtime. And as result: regenerator-runtime is not marked as external dependency and it becomes part of the package.

**Comment by JMarkoski** [comment]
@dangreen I'm not sure I understood correctly what is the problem you are trying to solve, but I'll try to explain the philosophy of using the `@babel/preset-env` with `useBuiltIns: 'usage'` and `@babel/transform-runtime`.

So, we should note the main important difference between the two: you use `@babel/preset-env` with `{useBuiltIns: 'usage', corejs: *}` when you want **globally polluting** polyfilling, using the babel built-ins for that, and you use `@babel/transform-runtime` when you want to **bundle** the polyfills with your app, so you don't pollute the global environment.

Now, having that philosophy in mind about the above mentioned usage of `@babel/preset-env`, the polyfilling process also includes the `regenerator-runtime/runtime` polyfill, together with the babel helpers like `classCallCheck` and the feature polyfills from the `core-js` library. In other words, with this config you say: 

> I want globally polluting polyfilling with babel built-ins that are included on every usage of the polyfillable feature.

Now, when you include the `@babel/transform-runtime` in the context, you are somewhat doing the same stuff but differently. First you said that you want globally polluting polyfills for everything, and now you say, I want polyfills/helpers for everything bundled with my app, not polluting the global environment. 

Actually you want to have this scenario, which I will explain why @nicolo-ribaudo says here that it doesn't make much sense: [https://github.com/babel/babel/pull/10768#pullrequestreview-324863979](https://github.com/babel/babel/pull/10768#pullrequestreview-324863979):

When `@babel/preset-env` with `{useBuiltIns: 'usage', corejs: *}` or `@babel/transform-runtime` is used, each one of them adds a **set** of things: 

1. Helpers like `classCallCheck`
2. Polyfills from `core-js`
3. Instance of `regenerator-runtime/runtime`

In the case of `@babel/preset-env` this is globally polluting, and in the case of `@babel/transform-runtime` is not, it adds references to the `@babel/runtime` or `@babel/runtime-corejs*`. But the important note is:

> The set is included as a whole.

You can't add globally polluting polyfills from `core-js` and add `regenerator-runtime/runtime` that is not globally polluting. You can't pick things from both sets and mix them, because how something includes/uses other thing is connected, and AFAIK mixing will lead to some unpredictability, or is hard to reason about. So, as @nicolo-ribaudo  says in the comment, it doesn't even make sense to mix globally polluting with globally safe pollyfilling, you either are fully globally polluting or fully globally safe.

This is my limited understanding of the matter, @nicolo-ribaudo feel free to correct me if I am wrong somewhere.

**Comment by dangreen** [comment]
@JMarkoski ok. How I should configure preset-env and transform-runtime to not pollute the global environment?

**Comment by dangreen** [comment]
@JMarkoski I should disable useBuiltIns for preset-env and add corejs option to transform-runtime?

**Comment by JMarkoski** [comment]
@dangreen Copy-paste from my answer given in (linked in the above comment) : [https://github.com/babel/babel/issues/9853#issuecomment-619587386](https://github.com/babel/babel/issues/9853#issuecomment-619587386)

> **Library**: If you are authoring a library, use only `@babel/transform-runtime` with `corejs` option plus `@babel/runtime-corejs3` as dependency. Also I would transpile packages I would use from `node_modules`. For this you will need to set the `absoluteRuntime` option ([https://babeljs.io/docs/en/babel-plugin-transform-runtime#absoluteruntime](https://babeljs.io/docs/en/babel-plugin-transform-runtime#absoluteruntime)) to resolve the runtime dependency from a single place, because `@babel/transform-runtime` imports from `@babel/runtime-corejs3` directly, but that only works if `@babel/runtime-corejs3` is in the `node_modules` of the file that is being compiled.

Note that, if you want to target specific environments with this, you can't do that, because `@babel/transform-runtime` doesn't have an option to specify targets, you'll get polyfills for every polyfillable feature.

**Comment by dangreen** [comment]
@JMarkoski in your answer no word about preset-env. I want to know how to combine preset-env with transform-runtime.

**Comment by JMarkoski** [comment]
@dangreen I don't seem to understand your question, may you please be more specific? According to this comment:

> @JMarkoski ok. How I should configure preset-env and transform-runtime to not pollute the global environment?

you don't want global pollution. Using `@babel/preset-env` includes polyfills by polluting the global environment. You can't configure it otherwise, it's what it does, includes imports from `core-js` that will pollute your global environment. The only option to avoid that is to use `@babel/transform-runtime`.

**Comment by dangreen** [comment]
@JMarkoski can you please provide babel config example for that?

Because config like that:

```
{
    "exclude": "node_modules/**",
    "presets": [
        ["@babel/preset-env", {
            "useBuiltIns": "usage",
            "corejs": 3
        }]
    ],
    "plugins": [
        ["@babel/plugin-transform-runtime", {
            "corejs": 3
        }]
    ] 
}
```

produce code like that:

```
import 'core-js/modules/es.regexp.exec';
import 'core-js/modules/es.string.match';
import _startsWithInstanceProperty from '@babel/runtime-corejs3/core-js-stable/instance/starts-with';
import _slicedToArray from '@babel/runtime-corejs3/helpers/esm/slicedToArray';
import _regeneratorRuntime from '@babel/runtime-corejs3/regenerator';
import 'regenerator-runtime/runtime';
```

**Comment by JMarkoski** [comment]
@dangreen: Make this:

> `npm install --save-dev @babel/plugin-transform-runtime`
> `npm install --save @babel/runtime-corejs3`

With this config:

```
{
    "exclude": "node_modules/**",
    "plugins": [
        ["@babel/plugin-transform-runtime", {
            "corejs": 3
        }]
    ] 
}
```

**Comment by dangreen** [comment]
@JMarkoski where is preset-env config?

**Comment by JMarkoski** [comment]
@dangreen You don't need it, moreover, you shouldn't use it with `@babel/transform-runtime` (although you can use it in a config where `useBuiltIns: 'entry'` and `@babel/transform-runtime` is used only for helpers, but here we talk about `useBuiltIns: 'usage'`). I'll link again my answer that answers exactly the question why you **shouldn't** use it: [https://github.com/babel/babel/issues/9853#issuecomment-619587386](https://github.com/babel/babel/issues/9853#issuecomment-619587386)

**Comment by dangreen** [comment]
@JMarkoski 

> you shouldn't use it with @babel/transform-runtime

to my mind this important info should be added to transform-runtime documentation

**Comment by JMarkoski** [comment]
@dangreen See the corrected comments. Hope it solves your problem :)

**Comment by dangreen** [comment]
@JMarkoski but, as I understand, transform-runtime doesn't respect targets and browserslist config, and transform-runtime can add useless polyfill. I'm right?

**Comment by JMarkoski** [comment]
Yes, that's true. But the benefit is that you will have the polyfills bundled with your library/app (its more useful for libraries, not so much for apps, because you can allow global pollution in your app, the problem is that you don't want your library users to have global scope pollution), and avoid global scope pollution.

The main thing and the main point of confusion is how to handle polyfills, i.e mixing `@babel/preset-env` with `useBuiltIns` with `@babel/transform-runtime`. You can remember it like this: Whenever you use `@babel/preset-env` for polyfill handling you pollute, whenever you use `@babel/transform-runtime` for polyfill handling you are safe, although including useless polyfills.

**Comment by nyngwang** [comment]
> [[source]](https://github.com/babel/babel/issues/10759#issuecomment-626209728)
> you should mention that this would lead to a less efficient bundles due to the addition of **extraneous polyfills**, that are not actually used by the library. However, it could work in certain scenarios.
> 
> [[source]](https://github.com/babel/babel/issues/10759#issuecomment-626175264)
> No, it won't. Adding import to `core-js` will pollute the global environment, and the single line import will convert to more granular imports from the `core-js` library **based on your targets**, thus not adding polyfills that are not needed.

@nicolo-ribaudo (Sorry for tagging, but I'm reviewing long-standing issues on Babel.)

I don't think this is correct but, I need to double-check with you as I saw your thumb-up there.

- When the `targets` is set to the last 3 versions, I've tested that `useBuiltIns: 'entry'` increases the bundle size. `@babel/plugin-transform-runtime` (or the shorthand `@babel/transform-runtime`, just a kind note for beginners) does not respect the browser `targets`, otherwise the [README of babel-polyfills](https://github.com/babel/babel-polyfills/tree/main#:~:text=it%20wasn't%20possible) would be incorrect.

For me, these arguments are also wrong:

> [[]]()

**Comment by nyngwang** [comment]
I'm here to clarify everything that made me feel frustrated when I first came across this thread. Now I have understood better about Babel, I decided to help future readers by clarifying some possibly misleading ideas here.

---

<details>
<summary>Does <code>useBuiltIns: 'entry'</code> add unnecessary polyfills?</summary>

> [[link to top]](https://github.com/babel/babel/issues/10759#issuecomment-626209728)
> JMarkoski you should mention that this would lead to a less efficient bundles due to the addition of extraneous polyfills, that are not actually used by the library. However, it could work in certain scenarios.

> [[link to top]](https://github.com/babel/babel/issues/10759#issuecomment-626214209)
> No, it won't. Adding import to `core-js` will pollute the global environment, and the single line import will convert to more granular imports from the `core-js` library **based on your targets**, thus **_not adding polyfills that are not needed_**.

All polyfills-related features under `@babel/preset-env` respect your browser targets, so this is correct. For reference, see: #14443. The confusing part is the line following it:

> , thus **_not adding polyfills that are not needed_**.

They try to convey that all those features a target browser does not _support_ are needed. So it's not about "usage". But adding **unused** polyfills is equivalent to adding something not needed from our perspectives.
</details>

---

<details>
<summary>Is <code>@babel/transform-runtime</code> only for polyfills?</summary>

Now, the interesting part:

> [[link to top]](https://github.com/babel/babel/issues/10759#issuecomment-627892271)
> [...] and you use `@babel/transform-runtime` when you want to bundle the polyfills with your app

From this comment, I'm afraid that JMarkoski might convey the wrong idea that the purpose of `@babel/transform-runtime` _**were**_ solely for polyfills by their wording. But the truth is: One can always opt out of the polyfills feature it provides, and use it to extract helper functions only.

This is why JMarkoski cannot understand the simple and valid question asked by dangreen:

> [[link to top]](https://github.com/babel/babel/issues/10759#issuecomment-627917241)
> ok. How I should configure preset-env and transform-runtime to not pollute the global environment?

> [[link to top]](https://github.com/babel/babel/issues/10759#issuecomment-628066036)
> dangreen I don't seem to understand your question, may you please be more specific? According to this comment: [...]

The question is simple and specific. We can use `preset-env` for transpiling syntax only and `transform-runtime` for adding pure polyfills only:

```js
plugins: [
  // `corejs` option enables the pure polyfills.
  ['@babel/plugin-transform-runtime', { corejs: 3 } ],
],
presets: [
  // disable polyfills from `preset-env`.
  ['@babel/preset-env', { useBuiltIns: false } ],
]
```

</details>

---


I think this thread can be closed as `regenerator` option has been removed from `@babel/plugin-transform-runtime` in #16310, and removing _unused_ imports like this might be problematic when people want to have **side effects**. (Kindly ping @nicolo-ribaudo)
